compiler/resolution/initializerRules.cpp

```diff
@@ -0,0 +1,226 @@
+                  // supported and old style constructors are deprecated
+                  // (Lydia note, 08/19/16)
+                  USR_FATAL(call, "can't make init call on type with old constructor style");
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+static
+void reorganizeBody(FnSymbol* fn, BlockStmt* phase1, BlockStmt* phase2,
+                    BlockStmt* otherInit);
+static
+void phase1Analysis(AggregateType* t, BlockStmt* phase1, BlockStmt* phase2,
+                    BlockStmt* otherInit);
+
+
+
+
+void handleInitializerRules(FnSymbol* fn, AggregateType* t) {
+  BlockStmt* phase1 = fn->body;
+  BlockStmt* phase2 = new BlockStmt();
+  BlockStmt* otherInit = new BlockStmt();
+
+  reorganizeBody(fn, phase1, phase2, otherInit);
+
+  phase1->insertAtTail(otherInit);
+  phase1->insertAtTail(phase2);
+
+  phase1Analysis(t, phase1, phase2, otherInit);
+}
+
+// This function traverses the body of the initializer backwards, moving the
+// statements it finds into the phase2 block statement until it encounters the
+// super/this.init() call or the start of the body, whichever comes first.
+// It then moves the super/this.init() call and the statements it relies on
+// into the otherInit block statement.
+static
+void reorganizeBody(FnSymbol* fn, BlockStmt* phase1, BlockStmt* phase2,
+                    BlockStmt* otherInit) {
+  while (phase1->body.length > 0) {
+    // Note - to make the default for an initializer body be phase 1,
+    // reverse the traversal order and perform some swaps of which block
+    // statement is receiving the code.  Will also need to update the creation
+    // of the phase1 block at the callsite, likely.
+
+    if (CallExpr* call = toCallExpr(phase1->body.tail)) {
+      if (CallExpr* inner = toCallExpr(call->baseExpr)) {
+        if (inner->isNamed(".")) {
+          if (SymExpr* sym = toSymExpr(inner->get(2))) {
+            if (VarSymbol* var = toVarSymbol(sym->var)) {
+              if (var->immediate->const_kind == CONST_KIND_STRING) {
+                if (!strcmp(var->immediate->v_string, "init")) {
+                  // While going backwards, we found the super/this.init() call
+                  // Time to stop moving into the phase2 block statement.
+                  otherInit->insertAtHead(call->remove());
+                  break;
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    // Behavior is not yet correct for super/this.init() calls within
+    // loops or if statements.  TODO: fix this
+
+    phase2->insertAtHead(phase1->body.tail->remove());
+  }
+}
+
+static
+void phase1Analysis(AggregateType* t, BlockStmt* phase1, BlockStmt* phase2,
+                    BlockStmt* otherInit) {
+  DefExpr* curField = toDefExpr(t->fields.head);
+  // the super field is always first
+  bool *seenField = (bool*)calloc(t->fields.length, sizeof(bool));
+  if (curField)
```

Comment:
It would mean either that we have no fields or that the fields vector differs from my expectations.